### PR TITLE
`execute` is now retry-able.  

### PR DESCRIPTION
**Pull request by nvthongswansea**

### CODE REVIEW

File: request.go [hunk 1]
```diff
@@ -35,14 +40,14 @@ type RequestStatusProperties struct {
 
 //RequestError error of a request
 type RequestError struct {
-	StatusMessage string `json:"status"`
-	ErrorMessage  string `json:"message"`
-	StatusCode    int
+	Title       string `json:"title"`
```
Comment (itakouna) [inline request.go]: Plz, use `Message`.

File: request.go [hunk 2]
```diff
@@ -10,6 +10,11 @@ import (
 	"time"
 )
 
+const (
+	requestCheckTimeout            = 2 * time.Minute
+	requestServerErrorRetryTimeout = 1 * time.Minute
```
Comment (itakouna) [inline request.go]: It would be great if this can be passed as parameter in the client's configuration.

File: request.go (hunk 1 above)
Comment (nvthongswansea) [reply request.go]: I think it is inconsistent with "title" JSON property if we name `Message`.

File: request.go (hunk 2 above)
Comment (nvthongswansea) [reply request.go]: OK, I'll implement it.

Comment (nvthongswansea) [reply request.go]: @itakouna just `requestServerErrorRetryTimeout `, isn't it?

File: request.go [hunk 3]
```diff
@@ -73,29 +78,55 @@ func (r *Request) execute(c Client, output interface{}) error {
...
+			c.cfg.logger.Debugf("Status code returned: %v", result.StatusCode)
+
+			if result.StatusCode >= 300 {
+				var errorMessage RequestError //error messages have a different structure, so they are read with a different struct
+				errorMessage.StatusCode = result.StatusCode
+				json.Unmarshal(iostream, &errorMessage)
+				if result.StatusCode >= 500 {
+					latestRetryErr = errorMessage
+					c.cfg.logger.Errorf("RETRY! Error message: %v. Title: %v. Code: %v.", errorMessage.Description, errorMessage.Title, errorMessage.StatusCode)
+					time.Sleep(500 * time.Millisecond) //delay the request, so we don't do too many requests to the server
```
Comment (itakouna) [inline request.go]: I would also expose a parameter `maximumNumberOfRetries` compute the waiting delay in between the retries.
`waitingDelay := requestServerErrorRetryTimeout/ maximumNumberOfRetries`

File: request.go (hunk 1 above)
Comment (itakouna) [reply request.go]: Ok. then, ignore my comment.

File: request.go (hunk 3 above)
Comment (nvthongswansea) [reply request.go]: I would say either `timeout` or `maxNumberRetries` is chosen for better practice, not both.

Comment (itakouna) [reply request.go]: quote from AWS `In addition to simple retries, each AWS SDK implements exponential backoff algorithm for better flow control. The idea behind exponential backoff is to use progressively longer waits between retries for consecutive error responses. You should implement a maximum delay interval, as well as a maximum number of retries. The maximum delay interval and maximum number of retries are not necessarily fixed values, and should be set based on the operation being performed, as well as other local factors, such as network latency.

Comment (nvthongswansea) [reply request.go]: That is exactly what I stated. Either `timeout` or `maxNumberRetries` is chosen. That means we can use either `timeout` with `delayInterval` or `maxNumberRetries` with `delayInterval`. we should not use `timeout`, `maxNumberRetries` and `delayInterval` at the same time, which is kind of a bad practice.

Comment (nvthongswansea) [reply request.go]: If you want this implementation `waitingDelay := requestServerErrorRetryTimeout/ maximumNumberOfRetries`. We still have to hardcode `maximumNumberOfRetries`.

File: config.go [hunk 4]
```diff
@@ -1,27 +1,35 @@
...
 	"os"
 	"runtime"
-
-	"github.com/sirupsen/logrus"
+	"time"
 )
 
 const version = "1.0.0"
 
+const (
```
Comment (itakouna) [inline config.go]: would u add `version = "1.0.0"` into the it?

File: request.go (hunk 3 above)
Comment (itakouna) [reply request.go]: Ok. let's use `maxNumberRetries` with `delayInterval`

**Pull request merged**